# Descriptive Statistics Review

## Before you start:

- Read the README.md file
- Comment as much as you can
- Happy learning!

## Context

![img](./diamonds.jpg)

In this lab we are going to work with data to understand the characteristics of a diamond that are most likely to influence its price. In this first part of the lab, we will explore and clean our data. 

The dataset we will be using is comprised of approximately 54k rows and 11 different columns. As always, a row represents a single observation (in this case a diamond) and each of the columns represent a different feature of a diamond.

The following codebook was provided together with the dataset to clarify what each column represents:


| Column  | Description  |
|---|---|
| Price  | Price in US dollars (326-18,823)  |
| Carat  | Weight of the diamond (0.2--5.01)  |
| Cut  | Quality of the cut (Fair, Good, Very Good, Premium, Ideal)  |
| Color  | Diamond colour, from J (worst) to D (best)  |
| Clarity  | A measurement of how clear the diamond is (I1 (worst), SI2, SI1, VS2, VS1, VVS2, VVS1, IF (best))   |
| x  | Length in mm (0--10.74)  |
| y  | Width in mm (0--58.9)  |
| z  | Depth in mm (0--31.8)  |
| Depth  | Total depth percentage = z / mean(x, y) = 2 * z / (x + y) (43--79)  |
| Table  | Width of top of diamond relative to widest point (43--95)  |

## Libraries
Pandas and numpy will be needed for the analysis of the data.

In [1]:
# Importing the libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

First import the data from the .csv file provided and assign it to a variable named `diamonds` and **drop the column with the index.**

In [2]:
# Importing the dataset
diamonds = pd.read_csv('diamonds.csv')

# Dropping this column which had the same info as the index
diamonds.drop(columns = ['Unnamed: 0'], inplace = True)

# Taking a look into the data after the drop
#diamonds.head()

In [3]:
# Checking the columns we have
diamonds.columns

Index(['carat', 'cut', 'color', 'clarity', 'depth', 'table', 'price', 'x', 'y',
       'z'],
      dtype='object')

In [4]:
# Testing the new library Gabi found! This looks amazing
# For large datasets, like this maybe, since it took like 1-2 mins to display, might want to try using
# profile = ProfileReport(diamonds, minimal = True)

from pandas_profiling import ProfileReport
profile = ProfileReport(diamonds, minimal = True)
profile

# 1. Taking the first look at the data.
Let's see how the data looks by using pandas methods like `head()`, `info()` and `describe()`. 

**First, use the `head` method.**

In [5]:
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

We can see the first 5 rows of the dataset using the `head` method. This by itself doesn't tell us much about the data that we have, but we can have a first look at the features (columns) and some of the values that each one takes.

**What do you see? Make some comments about the values you see in each column, comparing them with the codebook. Is that what you would expect for these variables?**

In [6]:
'''
Yes, the rows are what I would expect. They also seem to have the correct data types as well.
'''
# Checking the data types
diamonds.dtypes

carat      float64
cut         object
color       object
clarity     object
depth      float64
table      float64
price        int64
x          float64
y          float64
z          float64
dtype: object

It is very important to know the amount of data we have, because everything will depend on that, from the quality of the analysis to the choice of our infrastracture.

**Check the shape of the data**

In [7]:
diamonds.shape

(53940, 10)

The `clarity` column is confusing because we are not diamond experts. Let's modify this column with a new scale that is more understandable for us.

**Modify `clarity` column with numbers from 0 to 7. The lowest would be 0 with value `I1` and the greatest 7 with value `IF`**

In [8]:
# A measurement of how clear the diamond is (I1 (worst), SI2, SI1, VS2, VS1, VVS2, VVS1, IF (best))
# Doing the replacements
diamonds['clarity'].replace({'I1': 0, 'SI2': 1, 'SI1': 2, 'VS2': 3, 'VS1': 4,
                             'VVS2': 5, 'VVS1': 6, 'IF': 7}, inplace = True)
# Checking the result
diamonds.head()

carat      cut color  clarity  depth  table  price     x     y     z
0   0.23    Ideal     E        1   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E        2   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E        4   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I        3   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J        1   63.3   58.0    335  4.34  4.35  2.75

It makes sense to do the same with the `color` column.

**Do the same with values from 0 to 6. Read the codebook to see the match**

In [9]:
# Diamond colour, from J (worst) to D (best)
# Doing the replacements
diamonds['color'].replace({'D': 6, 'E': 5, 'F': 4, 'G': 3, 'H': 2, 'I': 1, 'J': 0}, inplace = True)

# Checking the result
diamonds.head()

carat      cut  color  clarity  depth  table  price     x     y     z
0   0.23    Ideal      5        1   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium      5        2   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good      5        4   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium      1        3   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good      0        1   63.3   58.0    335  4.34  4.35  2.75

With the `info` method, we can see the features of the dataset, and the amount of observations (rows) that have a non-null value and the types of the features. 

**Now use the `info` method and comparing with the shape, comment on what you see**

In [10]:
diamonds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
carat      53940 non-null float64
cut        53940 non-null object
color      53940 non-null int64
clarity    53940 non-null int64
depth      53940 non-null float64
table      53940 non-null float64
price      53940 non-null int64
x          53940 non-null float64
y          53940 non-null float64
z          53940 non-null float64
dtypes: float64(6), int64(3), object(1)
memory usage: 4.1+ MB


In the last line of the info output, you have some information about the types of the columns. As you know, it is a good idea to check if the types of each column is what you expect. If a column has the right type, we will be able to do all the operations that we want to do. 

For instance, if we have a column that is a `date` with a `string` format, we will have the data but we won't be able to do a simple operation, such as format the date the way that we would like.

Changing the data type to the one we need can help us to solve a lot of problems in our data.

**Check the types of each column and comment if it matches with the expected**

In [11]:
'''
Just like with the info() output above, the columns seem to have the desired format, have no nulls
and everything looks good. Don't know what else to say.
'''
diamonds.dtypes

carat      float64
cut         object
color        int64
clarity      int64
depth      float64
table      float64
price        int64
x          float64
y          float64
z          float64
dtype: object

# 2. A deeper look: checking the basic statistics.

The `describe` method gives us an overview of our data. From here we can see all the descriptive metrics for our variables.

**Use the `describe` method and comment on what you see**

In [12]:
'''
I see all the useful information describe gives us, like the count, mean, std and the quartiles.
Seems like the color and clarity mean are around 3, so the diamonds overall aren't that amazing.
Also, it is strange the fact that x, y and z have a min of 0, like the comments below mention.
'''
diamonds.describe()#.round(decimals = 2) # just a test to get rid of all these 0's, and it worked

carat         color       clarity         depth         table  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean       0.797940      3.405803      3.051020     61.749405     57.457184   
std        0.474011      1.701105      1.647136      1.432621      2.234491   
min        0.200000      0.000000      0.000000     43.000000     43.000000   
25%        0.400000      2.000000      2.000000     61.000000     56.000000   
50%        0.700000      3.000000      3.000000     61.800000     57.000000   
75%        1.040000      5.000000      4.000000     62.500000     59.000000   
max        5.010000      6.000000      7.000000     79.000000     95.000000   

              price             x             y             z  
count  53940.000000  53940.000000  53940.000000  53940.000000  
mean    3932.799722      5.731157      5.734526      3.538734  
std     3989.439738      1.121761      1.142135      0.705699  
min      326.000000      0.000000      0.000000      0.000000  
25%      950.000000      4.710000      4.720000      2.910000  
50%     2401.000000      5.700000      5.710000      3.530000  
75%     5324.250000      6.540000      6.540000      4.040000  
max    18823.000000     10.740000     58.900000     31.800000

You have probably noticed that the columns x, y and z have a minimum value of 0. This means that there are one or more rows (or observations) in our dataset that are supposedly representing a diamond that has lenght, width or depth of 0. Considering that we're talking about a physical object, this is impossible!

Now let's proceed to check the rows that have a value of 0 in any of the x, y or z columns. By doing this we want to check if the data we are missing can be obtained using the data that we do have.

**Check the columns with `x`, `y` and `z` with value 0 in all of them and comment what you see**

In [13]:
# Doing some filtering to check the results
x_0 = diamonds[diamonds['x'] == 0]#.count()
y_0 = diamonds[diamonds['y'] == 0]#.count()
z_0 = diamonds[diamonds['z'] == 0]#.count()

# Ok, this one seems to be working correctly. I was afraid it wasn't giving me back the DF with
# all of these values, but it seems to be doing it right
x_y_z_filter = diamonds[(diamonds['x'] == 0) | (diamonds['y'] == 0) | (diamonds['z'] == 0)]
x_y_z_filter#.count()
#x_0
#y_0
#z_0

carat        cut  color  clarity  depth  table  price     x     y    z
2207    1.00    Premium      3        1   59.1   59.0   3142  6.55  6.48  0.0
2314    1.01    Premium      2        0   58.1   59.0   3167  6.66  6.60  0.0
4791    1.10    Premium      3        1   63.0   59.0   3696  6.50  6.47  0.0
5471    1.01    Premium      4        1   59.2   58.0   3837  6.50  6.47  0.0
10167   1.50       Good      3        0   64.0   61.0   4731  7.15  7.04  0.0
11182   1.07      Ideal      4        1   61.6   56.0   4954  0.00  6.62  0.0
11963   1.00  Very Good      2        3   63.3   53.0   5139  0.00  0.00  0.0
13601   1.15      Ideal      3        3   59.2   56.0   5564  6.88  6.83  0.0
15951   1.14       Fair      3        4   57.5   67.0   6381  0.00  0.00  0.0
24394   2.18    Premium      2        1   59.4   61.0  12631  8.49  8.45  0.0
24520   1.56      Ideal      3        3   62.2   54.0  12800  0.00  0.00  0.0
26123   2.25    Premium      1        2   61.3   58.0  15397  8.52  8.42  0.0
26243   1.20    Premium      6        6   62.1   59.0  15686  0.00  0.00  0.0
27112   2.20    Premium      2        2   61.2   59.0  17265  8.42  8.37  0.0
27429   2.25    Premium      2        1   62.8   59.0  18034  0.00  0.00  0.0
27503   2.02    Premium      2        3   62.7   53.0  18207  8.02  7.95  0.0
27739   2.80       Good      3        1   63.8   58.0  18788  8.90  8.85  0.0
49556   0.71       Good      4        1   64.1   60.0   2130  0.00  0.00  0.0
49557   0.71       Good      4        1   64.1   60.0   2130  0.00  0.00  0.0
51506   1.12    Premium      3        0   60.4   59.0   2383  6.71  6.67  0.0

As you can see, we have 20 rows that have a value of 0 in some or all the aforementioned columns.
Most of them (12) are missing the z value, which we can obtain using the columns depth, x and y. 

20 rows with issues represent just 0.03% of our data (20 out of 53940) so it wouldn't be a big deal to remove them. Still, lets try to keep all the data we have. 

For those 12 rows, we will create a function that applies the formula given in the codebook and get the value of z.

**Create a function named `calculate_z` that applies the function in the codebook to one single row you give to the function**

In [14]:
# Total depth percentage = z / mean(x, y) = 2 * z / (x + y) (43--79)
# z = depth(x+y)/2
'''
Well... this was a hard one. Everyone was stuck in this for a while. After we all came up with
different solutions I tried adapting the first one I started working on with Ze's solution which
was the closest to what I was trying to do. Then I just tried changing his code a little bit
so I could understand it better.
For instance, I wasn't aware of this df.y being the same as df['y'] and it was making me kinda
confused
'''
def calculate_z(row):
    row["z"] = row["depth"] * (row["x"] + row["y"]) / 200
    return row


'''
Ze's code from which I changed a bit to get to the one I have above:
def calc_z(something):
        z = round(((something.x + something.y) / 2) * something.depth / 100, 2)
        return z

data.z = data.apply(lambda row: calc_z(row) if row.z == 0 else row.z , axis = 1)
'''

"\nZe's code from which I changed a bit to get to the one I have above:\ndef calc_z(something):\n        z = round(((something.x + something.y) / 2) * something.depth / 100, 2)\n        return z\n\ndata.z = data.apply(lambda row: calc_z(row) if row.z == 0 else row.z , axis = 1)\n"

**Apply it just to the rows with incorrect values**

In [15]:
diamonds = diamonds.apply(lambda row: calculate_z(row) if row.z == 0 else row, axis = 1)
diamonds

carat        cut  color  clarity  depth  table  price     x     y     z
0       0.23      Ideal      5        1   61.5   55.0    326  3.95  3.98  2.43
1       0.21    Premium      5        2   59.8   61.0    326  3.89  3.84  2.31
2       0.23       Good      5        4   56.9   65.0    327  4.05  4.07  2.31
3       0.29    Premium      1        3   62.4   58.0    334  4.20  4.23  2.63
4       0.31       Good      0        1   63.3   58.0    335  4.34  4.35  2.75
...      ...        ...    ...      ...    ...    ...    ...   ...   ...   ...
53935   0.72      Ideal      6        2   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72       Good      6        2   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70  Very Good      6        2   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    Premium      2        1   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75      Ideal      6        1   62.2   55.0   2757  5.83  5.87  3.64

[53940 rows x 10 columns]

If we leave the other 8 values as they are, it would negatively affect our analysis, because these are data that do not make logical sense. Therefore it is better to consider those values as NaN values, since they are probably the result of a mistake or error during process of measuring and storing these values in a dataset.

To replace them we can use the pandas .replace() method and np.NaN.

**Replace the zero values in the `z` column for a NaN**

In [16]:
# Replacing the 0's with NaN's
diamonds['x'].replace(0, np.nan, inplace = True)
diamonds['y'].replace(0, np.nan, inplace = True)
diamonds['z'].replace(0, np.nan, inplace = True)

In [17]:
# Checking it if worked
diamonds[(diamonds["x"] == 0) | (diamonds["y"] == 0) | (diamonds["z"] == 0)]

Empty DataFrame
Columns: [carat, cut, color, clarity, depth, table, price, x, y, z]
Index: []

Let's check the data again with the `describe()` method.

In [18]:
diamonds.describe()

carat         color       clarity         depth         table  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean       0.797940      3.405803      3.051020     61.749405     57.457184   
std        0.474011      1.701105      1.647136      1.432621      2.234491   
min        0.200000      0.000000      0.000000     43.000000     43.000000   
25%        0.400000      2.000000      2.000000     61.000000     56.000000   
50%        0.700000      3.000000      3.000000     61.800000     57.000000   
75%        1.040000      5.000000      4.000000     62.500000     59.000000   
max        5.010000      6.000000      7.000000     79.000000     95.000000   

              price             x             y             z  
count  53940.000000  53932.000000  53933.000000  53933.000000  
mean    3932.799722      5.732007      5.735270      3.540238  
std     3989.439738      1.119670      1.140339      0.702690  
min      326.000000      3.730000      3.680000      1.070000  
25%      950.000000      4.710000      4.720000      2.910000  
50%     2401.000000      5.700000      5.710000      3.530000  
75%     5324.250000      6.540000      6.540000      4.040000  
max    18823.000000     10.740000     58.900000     31.800000

The minimum value for x, y and z should now be a positive number, as it should be for the physical measurements of an object.

Let's finish by checking for NaN values in the data. Since we introduced them ourselves using 'replace', we will surely find some, but there may be more that are unrelated to the x, y and z columns. Checking NaNs is a fundamental part of data cleaning and it's always better to do this kind of operations before proceeding with analysis.

**Check how many NaNs do you have, comment what you would do with those values, and then do so**

In [19]:
'''
I see we have 8, 7 and 7 NaN's in the x, y and z columns, respectively. Honestly, at this point,
after having spent hours trying to figure out that function to calculate z, I would just drop
these rows, as they represent almost nothing in this gigantic dataset.
'''
diamonds.isna().sum()

carat      0
cut        0
color      0
clarity    0
depth      0
table      0
price      0
x          8
y          7
z          7
dtype: int64

In [20]:
# Dropping the NaN's that we introduced ourselves
diamonds.dropna(inplace = True)

# Checking if it worked. And it did.
diamonds.isna().sum()

carat      0
cut        0
color      0
clarity    0
depth      0
table      0
price      0
x          0
y          0
z          0
dtype: int64

# 3. Checking for outliers
Now we are going to revisit the summary table to check for outliers.

**Use the `describe` method again and comment on what you see. After that, check if you have any outliers** 

In [21]:
diamonds.describe()

carat         color       clarity         depth         table  \
count  53932.000000  53932.000000  53932.000000  53932.000000  53932.000000   
mean       0.797879      3.405789      3.051101     61.749336     57.457029   
std        0.473986      1.701165      1.647109      1.432514      2.234064   
min        0.200000      0.000000      0.000000     43.000000     43.000000   
25%        0.400000      2.000000      2.000000     61.000000     56.000000   
50%        0.700000      3.000000      3.000000     61.800000     57.000000   
75%        1.040000      5.000000      4.000000     62.500000     59.000000   
max        5.010000      6.000000      7.000000     79.000000     95.000000   

              price             x             y             z  
count  53932.000000  53932.000000  53932.000000  53932.000000  
mean    3932.136079      5.732007      5.735254      3.540265  
std     3988.734835      1.119670      1.140343      0.702667  
min      326.000000      3.730000      3.680000      1.070000  
25%      949.750000      4.710000      4.720000      2.910000  
50%     2401.000000      5.700000      5.710000      3.530000  
75%     5324.000000      6.540000      6.540000      4.040000  
max    18823.000000     10.740000     58.900000     31.800000

In [ ]:
#your comments here

To manage these outliers, we are going to filter our DataFrame, we're going to take all the values that have a price higher than the 75th percentile.

**Look for that quantile and filter the dataframe to clearly see the outliers. What do you think?**

In [ ]:
#your code here

Our dataset is really big and the outliers are really far apart from the rest of the values. To see this more clearly we will use a boxplot, which plots the median, 25th and 75th quartile, the maximum and minimum, as well as any outliers.

In [ ]:
#Run this code
fig, ax = plt.subplots(1,2, figsize=(10, 5))
sns.boxplot(y=diamonds.y, ax=ax[0])
sns.boxplot(y=diamonds.z, ax=ax[1])
plt.subplots_adjust(wspace=0.5)

Now we can see that all the values are within an acceptable range, but we have 2 big outliers in y and 1 in z. Now we know that our max values for y should be around 10 and the values for z should be around 6, so let's filter our dataset to find values for z higher than 10 in it.


In [ ]:
#your code here

Now that we have found the outlier, let's use the function we defined earlier to correct this value. First, we need to change the value to 0 (because that's how we defined the function before) and then we will apply it.

**Apply `calculate_z` for the row with the outlier**

In [ ]:
#your code here

Let's check if we actually corrected the outlier.

In [ ]:
diamonds.loc[48410]

Now let's do the same for `y`. First, let's filter the DataFrame to find the outliers. We said that the maximum values should be around 10, so let's check what are the values above 10.

**Check the values greater than 10 in the `y` column** 

In [ ]:
#your code here

We can clearly see that the 31.8 in row 49189 is an outlier for the y value. Also, we can see that the 58.9 value for `y` in row 24067 is actually its depth, so it was a mistake when they introduced the data. Let's create a function to fix these outliers.

**Create a function named `calculate_y` to calculate `y` using `z` and `x` the same way you did above**

In [ ]:
#your code here

We will check the rows that had an outlier in `y`, to check that the values were changed.

**Check those rows and comment what you see**

In [ ]:
#your thoughts here

Now that we have corrected or dropped all of our outliers, lets plot another box plot to double check.

In [ ]:
#Run this code
fig, ax = plt.subplots(1,2, figsize=(10, 5))
sns.boxplot(y=diamonds.y, ax=ax[0])
sns.boxplot(y=diamonds.z, ax=ax[1])
plt.subplots_adjust(wspace=0.5)

**What do you think? Are these values more reasonable?**


In [ ]:
#your thoughts here

**Once you are happy with your cleaning, save the cleaned data and continue to csv. Your new csv should be named ``diamonds_clean.csv``**

In [ ]:
#your code here